File: common/src/main/java/juuxel/adorn/util/animation/AnimatedPropertyWrapper.java
package juuxel.adorn.util.animation;

import java.util.function.Consumer;
import java.util.function.Supplier;

public final class AnimatedPropertyWrapper<T> extends AbstractAnimatedProperty<T> {
    private final Supplier<T> getter;
    private final Consumer<T> setter;

    public AnimatedPropertyWrapper(AnimationEngine engine, int duration, Interpolator<T> interpolator, Supplier<T> getter, Consumer<T> setter) {
        super(engine, duration, interpolator);
        this.getter = getter;
        this.setter = setter;
    }

    @Override
    public T get() {
        return getter.get();
    }

    @Override
    protected void setRawValue(T value) {
        setter.accept(value);
    }
}

File: common/src/main/kotlin/juuxel/adorn/util/animation/AnimatedPropertyWrapper.kt
package juuxel.adorn.util.animation

class AnimatedPropertyWrapper<T>(
    engine: AnimationEngine,
    duration: Int,
    interpolator: Interpolator<T>,
    private val getter: () -> T,
    private val setter: (T) -> Unit
) : AbstractAnimatedProperty<T>(engine, duration, interpolator) {
    override fun get(): T = getter()
    override fun setRawValue(value: T) = setter(value)
}
